Run some jdt.ui tests

datho7561 opened the pull request:
Fixes eclipse-jdtls/eclipse.jdt.javac#29

mickaelistria committed:
Make MatchLocator capable of working with DOM

Extra strategy in MatchLocator to use DOM, and tweaks in various
Locators to properly handle DOM. Tests in JavaSearchTests help a lot
(currently ~130/270 passing)

It basically clones and adapt some methods and adapt them to DOM,
A new flag is introduced to decide the strategy.

Done-ish (often with remaining issues)
* FieldLocator
* LocalVariableLocator
* MethodLocator
* SuperTypeReferenceLocator
* TypeDeclarationLocator
* TypeParameterLocator
* TypeReferenceLocator
TODO:
* AndLocator
* ModuleLocator
* PackageDeclarationLocator
* PackageReferenceLocator
* OrLocator
* VariableLocator

Make some match locator work with DOM

Fix NPE, convert 4 errors to fails

Signed-off-by: Rob Stryker <stryker@redhat.com>

Handle match files in the order they arrived

Signed-off-by: Rob Stryker <stryker@redhat.com>

Fix testBug251827b and 5 others - import declarations java element must be found

Signed-off-by: Rob Stryker <stryker@redhat.com>

Move the DOM-based search into the javac bundle

Signed-off-by: Rob Stryker <rob@oxbeef.net>

Cleanup

Signed-off-by: Rob Stryker <stryker@redhat.com>

Move the DOM-based search into the javac bundle

Signed-off-by: Rob Stryker <stryker@redhat.com>

datho7561 committed:
Hack to prevent OperationCanceledException from being propagated to logs

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Completion for method references in javadoc

- Fixes for the relevance numbers for methods in javadoc
- Suggest constructors (including the default constructor) in `@see` and
  `@link`
- Do not show completion when completing parameter name in method
  reference in Javadoc
- when referencing a method from Javadoc,
  and the cursor is in the method parameters,
  specifically right after a `(` or `,`,
  complete the entire method signature

This should fix 37 test cases

Signed-off-by: David Thompson <davthomp@redhat.com>

Commit:
Fix CompletionContextTests0123 inside JavacConverter

Signed-off-by: Rob Stryker <stryker@redhat.com>

akurtakov committed:
Record junit issues and warn on increase

mickaelistria committed:
Work on completion with static import

* Return more proposals
* Do no repeat extra import in proposals

Commit:
getDeclaredTypes() on a type binding loses the parameterized info for children

Signed-off-by: Rob Stryker <stryker@redhat.com>

mickaelistria committed:
More support for annotation value names

datho7561 committed:
Extension point to replace `CompletionEngine`

Instead of using the
`-DCompilationUnit.codeComplete.DOM_BASED_OPERATIONS=true`,
make an extension point that allows replacing the CompletionEngine in `Openable`.
With the extension point,
we can contribute DOMCompletionEngine as a part of our javac fragment,
in a similar way to how it's being done for
`JavacCompilationUnitResolver`.

Signed-off-by: David Thompson <davthomp@redhat.com>

Commit:
Mark malformed

Signed-off-by: Rob Stryker <stryker@redhat.com>

mickaelistria committed:
Improve support for completing enum constants in switch case

akurtakov committed:
Publish all issues in Jenkins

mickaelistria committed:
Fix possible ClassCastException trying to complete typeparam

datho7561 committed:
Collect `instanceof` and switch pattern variable bindings

Fixes #1130

Signed-off-by: David Thompson <davthomp@redhat.com>

akurtakov committed:
Fix unused import

As there are such cases often I raised it to error and enable the save
action.

akurtakov committed:
Obey SuppressWarnings("removal") in unused detector

Makes it slightly easier to use javac branch on big codebases as
hundreds of explicitly suppressed such warnings are now properly not
displayed anymore.

snjeza committed:
Unused imports are always listed as errors

snjeza committed:
Test case for PR #1139

Signed-off-by: Snjezana Peco <snjezana.peco@redhat.com>

Commit:
NPE in binding resolver

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
Fixes 1145 - don't make a position when length is invalid

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
Unnecessary import

Signed-off-by: Rob Stryker <stryker@redhat.com>

mickaelistria committed:
Improve converter & binding for IMemberValuePair

mickaelistria committed:
Set range for missing name anyway

properly set offset, with a length of 0.

mickaelistria committed:
Improve expectType for ArrayInitializer

datho7561 committed:
Completion for pattern variable initialized in for loop expression

eg. at the `|`:

```java
import java.util.List;
public class Main {
  private static record MyRecord(String a, int b) {}
  public static void main(String... args) {
    List myList = List.of(new MyRecord("hello", 12), new Object());
    for (int i = 0; i < myList.size() && myList.get(i) instanceof MyRecord(String jello, int kilometer); i++) {
      System.out.println(jel|);
    }
  }
}
```

Upstream doesn't handle this properly yet, so I added a test suite for
it.

Also handles collecing variables declared in the `for` initializer for
completion.

Signed-off-by: David Thompson <davthomp@redhat.com>

mickaelistria committed:
Improve some proceedOnError generation

mickaelistria committed:
Fix some relevance for member dereference

Commit:
Fixes #1155 - remove all ClassCastException in error log after importing org.eclipse.pde.ui

Signed-off-by: Rob Stryker <stryker@redhat.com>

Cleanup

Signed-off-by: Rob Stryker <stryker@redhat.com>

akurtakov committed:
Map compiler.warn.override.varargs.missing

akurtakov committed:
Ignore diagnostics without positions

"compiler.note.removal.plural.additional",
"compiler.note.unchecked.filename" don't have positions thus are not
usable in the ide.

akurtakov committed:
Fix failures of GHAs in local forks

As Javac needs Java 23 add it. While on it update Maven version used.

akurtakov committed:
Map one more varargs diagnostic

Cases like:
```
!ENTRY org.eclipse.jdt.core 4 0 2025-01-22 10:50:27.539
!MESSAGE Could not accurately convert diagnostic
(compiler.warn.varargs.unsafe.use.varargs.param)
/home/akurtakov/git/eclipse.platform.ui/tests/org.eclipse.jface.tests.databinding/src/org/eclipse/core/tests/databinding/observable/list/ListDiffTest.java:148:
warning: [varargs] Varargs method could cause heap pollution from
non-reifiable varargs parameter differences
		return Diffs.createListDiff(differences);
```
should be properly mapped now.

akurtakov committed:
Add api filters

In order to have javac branch with bigger version of jdt.core so it
always overrides it during testing.

mickaelistria committed:
Test for release

akurtakov committed:
Add forgotten api filters

In order to have javac branch with bigger version of jdt.core so it
always overrides it during testing.

akurtakov committed:
Enable api-check and javadoc profiles

Should reduce problems in workspace.

Commit:
Dogfooding the pde project has revealed several NPEs

Signed-off-by: Rob Stryker <stryker@redhat.com>

mickaelistria committed:
Use JavaTask in JavaCompiler

Using JavaTask allows a more proper initialization, for example with
better support for release settings.

Commit:
Fixes #1170 - missing diagnostic constant

Signed-off-by: Rob Stryker <stryker@redhat.com>

datho7561 committed:
Collect foreach variable for completion

- [ ] needs a test (there are no tests for this upstream)

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Fix completion at beginning of method

eg.

```java
public class Main {

  static int myMethod() {}

  public static void main(String... args) {
    System.out.println(myMethod(|));
  }

}
```

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Fix extends and implements

The type search wasn't running because there was no expected type.
...so nothing was suggested.

Also, we weren't properly filtering out records, enums, and sealed types
when suggesting classes to extend/implement.

Signed-off-by: David Thompson <davthomp@redhat.com>

mickaelistria committed:
Avoid NPE in JavacBindingResolver

mickaelistria committed:
Recover from CompletionFailure in JavacTypeBinding

Use the recover.isRecoveringBinding() flag to decide whether to return a
binding or not in case of CompletionFailure.

datho7561 committed:
Improve completion for case in switch that's not enum

eg.

```java
final int MY_NUM = 1;
final int MY_NUM2 = 12;
int i = 3;
switch (i) {
case MY_|
}

```

Currently, we were suggesting nothing,
since we assumed it was always enum.

Signed-off-by: David Thompson <davthomp@redhat.com>

Commit:
[Dogfooding] Fix issue where Record params are marked as unused

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
[Dogfooding] Fix unrecognized call to constructor to mark as used

Signed-off-by: Rob Stryker <stryker@redhat.com>

datho7561 committed:
Fix keyword token range for completion

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Do not suggest variables from current line for completion

This PR causes a regression: `CompletionTests10.test0007_not_inside_expression`.
I disagree with the results of this test under ECJ;
a type reference to `String` is suggested,
which is why there are 14 completion items.
I don't understand why `String` is suggested.
There used to be 14 completion items because we were suggesting `a`.

Signed-off-by: David Thompson <davthomp@redhat.com>

mickaelistria committed:
Do not suggest "extends" on enum or record

mickaelistria committed:
Avoid NPE in JavacBindingResolver (cont'd)

Fix condition

datho7561 committed:
Complete super constructor invocation

eg.

```java
public class MyClass extends MyOtherClass {
  public MyClass() {
    sup|
  }
}

class MyOtherClass {
  public MyOtherClass(int i) {}
  protected MyOtherClass(Object i) {}
  MyOtherClass(boolean b) {}
  private MyOtherClass(String s) {}
}
```

Requires #1206 in order to work properly if you have substring matching
enabled in your IDE.

Fixes #1205

Signed-off-by: David Thompson <davthomp@redhat.com>

Commit:
Move fix to UnusedTreeScanner

Signed-off-by: Rob Stryker <stryker@redhat.com>

datho7561 committed:
Set token range for completion proposals

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Add support for `MyClass.|this` completion

`ThisExpression` is it's own node, which we will need to handle
differently.

Should fix 2 test cases.

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Fix completing `out` from `System.|`

It used to insert `System.System.out`,
but now it properly inserts `System.out`.

This change causes no regressions,
and no new test passes.
I don't know what the code I deleted was supposed to do;
I'll figure that out before merging.

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Member types completion

eg.

```java
public class Main {

  public class Member {

  }

  public void method() {
    Main.Mem|
  }
}
```

Also handles:

```java
import java.util.Map;
public class Main implements Map.En| {
}
```

Signed-off-by: David Thompson <davthomp@redhat.com>

Commit:
Fixes #1186 (possibly) - exceptions have wrong name in method mouseover
popup box

Signed-off-by: Rob Stryker <stryker@redhat.com>
Also-By: Mickael Istria <mistria@redhat.com>

datho7561 committed:
Filter types based on access level

- suggest types in more cases
  (do not shut off type name completion if there are no expected types)
- do not suggest types when completing type arguments for a
  non-generic type (or if there are eg. 2 params and 3 args)
- do not suggest types from the default package in classes not in the
  default package
- do not suggest private inner classes
- only suggest package private classes when in the same package
- only suggest protected inner classes when the class the cursor is in
  inherits from the class containing the protected inner class

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Move `ExpectedTypes` into javac bundle

Fixes #1202

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
In type completion, filter by unique FQN

Should fix 6 cases without regressions,
because one of the projects used to test completion
has two classes called `SuperClass` in the
default package.

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Fix some bugs in completion type filtering

- Do prefix filtering first; string comparison should be fast compared
  to more involved operations
- Fix bad implementation of `protected` member classes;
  `getParent()` wasn't the right method to get the parent class of a
  member class
- Impossible classes filtering was wrong, it was checking methods
  instead

Signed-off-by: David Thompson <davthomp@redhat.com>

mickaelistria committed:
Avoid NPE in JavacBindingResolver

mickaelistria committed:
Improve some support for completing missing type

Failback to type of declaration from Search if not resolved

datho7561 committed:
Fix `TryStatement` always marked as malformed by `JavacConverter`

Fixes:
`ASTConverterTest.test0113`
`ASTConverterTest.test0114`
`ASTConverterTest.test0115`

as well as tests for other Java versions

Note: we only need to accomodate Java >= 8 in the AST conversion process,
since that's what upstream does now.

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Fixes to array dimensions in `JavacConverter`

Should fix 17 test cases

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Fix source range for array types when the `[` is unicode escape

eg. `MyType\u005b] myTypeArray = null;`

Fixes 4 tests:
`ASTConverterTest.test0266` and the versions for other Java compliances

Signed-off-by: David Thompson <davthomp@redhat.com>

akurtakov committed:
Stop dealing with JLS2

AST no longer supports it thus no need for Javac converter to be
complicated for nothing.

datho7561 committed:
Improve logic for 'inheritance' part of completion relevance

It's now taken into account for members of a qualified type being completed.

Signed-off-by: David Thompson <davthomp@redhat.com>

akurtakov committed:
Remove pre JLS 8 convertions

Not needed as neither Javac nor ECJ nor JDT makes use of anything pre
Java 8 nowadays.
VariableDeclaration.extraArrayDeclarataion is not supported in Java 8+
but was used without guards which shouldn't have happened.

datho7561 committed:
Improve extends/implements completion

- precalculate extends or implements info
- When there is no prefix, suggest types declared in the same class,
  eg. suggest `MyInterface` in the following snippet

```java
public class MyClass implements | {

}

interface MyInterface {}
```

- Better recovery of bindings for classes with incomplete extends or
  implements sections
  - previously the binding key was always `*`, which caused many that
    relied on the binding key to break

- Use ExtendsOrImplements info filtering in more places

Signed-off-by: David Thompson <davthomp@redhat.com>

mickaelistria committed:
Fix base conflict resolution

mickaelistria committed:
Improve completion for missing type

mickaelistria committed:
Enable debug output for failing mvn command

datho7561 committed:
Move key -> signature methods to `DOMCompletionUtil`

Fixes #1224

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Stabilize testDuplicateLocal

- Fix the number part of IVariableBinding's key
  - This means that the bindings for different variables with the same
    name in the same method can be distinguished eg. `rob` and `i` can
    be distinguished:

```java
public class Main {
	public void myMethod() {
		for (int i = 0; i < 10; i++) {
			int rob = 12;
		}
		for (String i = "asdf"; "hjkl".equals(i); i = "hjkl") {
			String rob = "rob";
		}
	}
}
```

- As a result, completion works deterministically in these cases
  (eg. before you just had to hope that the correct binding was cached
  in order to get the right results)

Signed-off-by: David Thompson <davthomp@redhat.com>

mickaelistria committed:
Fix more CompletionWithMissingTypeTests

mickaelistria committed:
Adapt to new API for JavacCompilerFactory

mickaelistria committed:
Empower JavacBindingResolver.resolveWellKnownType

Make it capable of resolving *any* type visible to this build.

mickaelistria committed:
Adapt conversion to new TypePattern setter

mickaelistria committed:
Do not check that some types are not returned by resolveWellKnownType

resolveWellKnownType method specifies a minimum set of supported types,
but the list is not meant to be exclusive and the definition of
"well-known" type is not so constraining.
So we allow test to pass even if a binding resolve can resolve
java.lang.Runnable as a well-known type.

mickaelistria committed:
Return binding for missing type too

+ do not add import for fully qualified types

mickaelistria committed:
Binding don't return irrelevant PackageBinding

+ better support for missing types (we reported as packages)

Commit:
Match locator was not initialized

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
JavaSearchTests.testCoreException - CoreException should be propagated, not caught

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
Pattern Wrappers must be cached to be reused

Signed-off-by: Rob Stryker <stryker@redhat.com>

datho7561 committed:
Fix some test cases for `this` completion

Signed-off-by: David Thompson <davthomp@redhat.com>

mickaelistria committed:
some more proceedOnError

datho7561 committed:
Casted receiver completion

eg.

```java
if (node instanceof MethodDeclaration) {
  node.getB|
}
```

completes to

```java
if (node instanceof MethodDeclaration) {
  ((MethodDeclaration)node).getBody()|
}
```

Fixes # 1242

I also took liberty in implementing/fixing:
- https://github.com/eclipse-jdt/eclipse.jdt.core/issues/3738
- https://github.com/eclipse-jdt/eclipse.jdt.core/issues/3737
- https://github.com/eclipse-jdt/eclipse.jdt.core/issues/3739
- https://github.com/eclipse-jdt/eclipse.jdt.core/issues/3740
- Overall doing a bit more flow analysis to suggest this completion in more cases

Overall, this causes two regressions which are related to limitations in
how JDT handles this type of completion.

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Qualify field accesses with `ClassName.this` when they are shadowed

- Handle suggesting anonymous class members
- Add a test for suggesting anonymous class members

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Change `@since` tag so that api tools doesn't complain

Signed-off-by: David Thompson <davthomp@redhat.com>

Commit:
[DOM Search] Fix 100+ tests

Signed-off-by: Rob Stryker <stryker@redhat.com>

mickaelistria committed:
Support for --limit-modules

snjeza committed:
Issues with lombok

Signed-off-by: Snjezana Peco <snjezana.peco@redhat.com>

mickaelistria committed:
Improve completion filtering for deprecation

mickaelistria committed:
Fix for .class completion with unknown type

datho7561 committed:
Completion for field modifiers

~53 new test passes

Fixes #1253

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Allow using any static fields when initializing field

See `CompletionTests.testBug332268a`

eg. should allow completing `val4` in the example below:

```java
public class Main {
  public int val1 = 12;
  public int val2 = 13;
  public int val3 = |;
  public static int val4 = 15;
}
```

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Set token range for completions built out of IType s

Should fix 1 case, and there are many cases that this helps to make less
wrong.

Signed-off-by: David Thompson <davthomp@redhat.com>

mickaelistria committed:
Get JavacTypeBinding.getDeclaredMethod ordered

a bit more reliably according to order in the class file (using JDT
model and its underlying ClassReader preserving order), so ExpectedTypes
can rely on it.

mickaelistria committed:
Don't filter deprecated nodes inside same unit

datho7561 committed:
Run search tests using javac

Uses a similar method to the existing borrowed tests:
add a new class to the javac test bundle that extends the upstream test suite.

Signed-off-by: David Thompson <davthomp@redhat.com>

mickaelistria committed:
Better resolve some variable

datho7561 committed:
Suggested name completion

Handle suggested names. eg the following suggests `myObject`:

```java
MyObject |
```

The following suggests `myObjects`:

```java
MyObject[] |
```

I also need to fix union type bindings as a part of this PR in order to
properly handle catch statements.

This PR DOESN'T handle variables that are referenced but not declared,
eg. `myVar` should be suggested here, but it isn't yet:

```java
int |
myVar = myVar + 12;
```

Fixes #1263

Signed-off-by: David Thompson <davthomp@redhat.com>

mickaelistria committed:
Detect expected types in lambda + treat boxed as native

mickaelistria committed:
Fix relevance for primitive/boxed type

mickaelistria committed:
Do not pass generated parameter names to method proposal

mickaelistria committed:
Fix completion for myMethod(|

Should suggest the methodRef

mickaelistria committed:
Fix some packages in CompletionTests18

datho7561 committed:
Complete names of undeclared but assigned variables

- do not erase bodies of methods in anonymous classes and class
  statements
  - this is over-permissive at this point,
    we'll need to refine it later

Signed-off-by: David Thompson <davthomp@redhat.com>

Commit:
[DOM Search] minor refactor

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
[DOM Search] - Fixes 14 search tests

Signed-off-by: Rob Stryker <stryker@redhat.com>

mickaelistria committed:
Fix relevance for method reference

mickaelistria committed:
Replace wildcard in proposal signatures

mickaelistria committed:
Fix some cases of methodreference

mickaelistria committed:
Explicitly avoid lint when no problems are to be reported

datho7561 committed:
Fix fine grain search with instanceof

Should fix 2 test cases,
these test cases were causing us XML-related headaches.

Signed-off-by: David Thompson <davthomp@redhat.com>

mickaelistria committed:
Temporarily disable issue to get build complete

mickaelistria committed:
Cache some JavaElement for bindings

Computing JavaElement can be expensive, cache result.

mickaelistria committed:
Ignore model when typeBinding.getDeclaredMethods() is from source

When from source, the order is fine, so don't bother with the model at
all and stick with Javac symbols

mickaelistria committed:
Cache key and simplify JavacMethodBinding.getJavaElement()

mickaelistria committed:
Fix method comparion (signature instead of name)

datho7561 committed:
Qualified completion in case statements

eg. completion at `|` should yield VALUEABLE_1 and properly filter out
eg. methods and non-static/final fields.

```java
public class DemoClass {
	void thisMethod() {
		int i = 12;
		switch (i) {
			case MyInner.|
		}
	}
	public static class MyInner {
		final static int VALUEABLE_1 = 1;
	}
}
```

Fixes 1 case, improves a few others so that only
the relevance numbers are wrong.

Signed-off-by: David Thompson <davthomp@redhat.com>

Commit:
[DOM Search] - Fix 6 tests related to generics

Signed-off-by: Rob Stryker <stryker@redhat.com>

datho7561 committed:
Filter out already included union types

eg. completion at `_` should not suggest `ExceptionA`

```java
try {
  throw new Exception();
} catch (ExceptionA | _ ) {
  // handle error
}
```

Fix some relevance constants related to this case.

Should fix at least 2 tests.

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Expected type for SingleVariableDeclaration in CatchClause

- consolidate logic for `CatchClause` into one place
- figure out which types are thrown in the body of the try statement,
  prioritize them
- figure out which types are already caught, deprioritize them
  ("uninteresting bindings")
- handle bug related to classes declared with type declaration statements
  - their bindings are still wrong, so that needs work

Signed-off-by: David Thompson <davthomp@redhat.com>

Commit:
[DOM Search] Solve 10 issues related to match rule

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
[DOM Search] Fix JavaSearchGenericTypeEquivalentTests.testTypeSingleArgument01 and others

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
[DOM Search] Fix 12 more issues

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
[DOM Search] Fix 10 issues in resolving match levels

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
NPE revealed when a test fails

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
Handle regressions in binding key generation

Signed-off-by: Rob Stryker <stryker@redhat.com>

Fix many regressions in previously successful tests

Signed-off-by: Rob Stryker <stryker@redhat.com>

Fix regression in org.eclipse.jdt.core.tests.dom.ASTConverterTestAST3_2 test0666

Signed-off-by: Rob Stryker <stryker@redhat.com>

Fix regressions in completion

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
[DOM] Fix testBug497719_0002, allowing super field and field access in try-with-resources

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
[DOM] implement illegal argument exception in binding.createArrayType(num)

Signed-off-by: Rob Stryker <stryker@redhat.com>

datho7561 committed:
Fix IType key sometimes doesn't end with `;` when it should

- Use `$` to separate inner class and outer class in IType key

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Update versions in parent POM references for javac bundle

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Hopefully fix the model test failure

Version range seems incorrect, I suspect that's the root cause

Signed-off-by: David Thompson <davthomp@redhat.com>

Commit:
Undo some changes in jdt.core that are no longer relevant

Signed-off-by: Rob Stryker <stryker@redhat.com>

mickaelistria committed:
Prevent CompletionFailure from bubbling up (and freezing IDE)

Reconciling or other routines may receive a CompletionFailure in some
-more or less valid- cases. Instead of letting the exception bubble up
to user land, we simply log it and continue anyway.

mickaelistria committed:
Small fix to CompletionEngines and ExplectedTypes

Should improve a couple tests
+ Added a Timer utility class to help profiling

datho7561 committed:
Suggest keywords before the top level type

eg.
```java
pub|
```

eg.
```java
package test;
import java.util.List;

abst|
```

- Detect that aren't connected to a type declaration using the text content
  of the CU
- Apply the rules regarding which modifiers prevent others from being
  used
- Prevent statement-like keywords in import or package declarations

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Completion for `case` keyword

eg.

```java
switch (i) {
  cas|
}
```

Signed-off-by: David Thompson <davthomp@redhat.com>

Commit:
[DOM Search] - Fixes testParameterizedTypeSingleArgument01 and 6 others

Signed-off-by: Rob Stryker <stryker@redhat.com>

datho7561 committed:
Completion for break and yield in switch statements and expr

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Keyword completion: differentiate between expressions and statements

- Add a bunch more statement-like keywords
- Handle completion for `default` in switch statements and expressions

Should fix around 43 cases

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Many keyword completion fixes

- Completion for `new`, `final`, `true`, `false`, and `null` keywords
- Improve logic for checking if only expression-like keywords should be
  suggested
- Prevent default completions in switches without a case label/default
  label

Signed-off-by: David Thompson <davthomp@redhat.com>

mickaelistria committed:
Lazily resolve DOMCompletionContext.getEnclosingElement()

Commit:
[DOM Search] Fix testDeclarationOfReferencedTypes05 and 1 others

Signed-off-by: Rob Stryker <stryker@redhat.com>

datho7561 committed:
Keyword completion fixes for try/catch/finally and do-while

- Disable default completion in completion after a try statement without
  a catch or finally
- Complete `catch` and `finally` after try statements
  (do not suggest `finally` if there's already a `finally` statement)
- Suggest `while` after a do-while statement if it's missing
- Prevent default completion after a do-while statement if it's missing
  the `while` keyword

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Another round of keyword completions

- Completion for `instanceof` keyword
  - Since it's sort of like an infix operator,
    I needed to write some more involved logic to get it to work properly.
- Suggest all type-declaration like keywords instead of just `class`
  - eg. `enum`, `interface`, `record`
- Complete the `import` keyword
  - See https://github.com/eclipse-jdt/eclipse.jdt.core/issues/3799,
    I won't fix that test case
- fix module import completion (only show it when `import module` is used)
- fix replace range on package completion
  - It uses the nested engine currently and the root cause was us not
    modifying the internal state correctly.
    I think using nested engine is bad,
    but I did the lazy hack instead of replacing the nested engine
    with our own `ISearchRequestor` implementation

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Improve "override method" completion

- Handle qualifying types referenced in the method declaration
  - eg. do not qualify imported types
  - eg. do not qualify types from the same package
  - eg. do not qualify inner classes nor inherited inner classes
- Fix parameter names
  - get them from the `IMethod`, which has more accurate
- Handle completion in anonymous classes
- Do not suggest overriding a method that's already overriden

Signed-off-by: David Thompson <davthomp@redhat.com>

Commit:
[DOM Search] implement matching on wildcards, extends, super, and other bugs

Signed-off-by: Rob Stryker <stryker@redhat.com>

datho7561 committed:
Cache type hierarchy in `DOMCompletionUtil.findInSupers`

Speeds up completion of types significantly,
since the type hierarchy of `java.lang.Exception` is only created once
every completion invocation.

Signed-off-by: David Thompson <davthomp@redhat.com>

mickaelistria committed:
Avoid calling getKey() too greedily

getKey can be very expensive, so we should only have it computed on
demand.
This saves ~30% when executing completion tests with Javac.

datho7561 committed:
Primitive types can autobox to `java.lang.Object`

(for expected types purposes)

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Prevent NPEs in completion tests

- Prevent NPE when completing name of the last parameter in a method
  declaration
  - This uses the existing "suggest a name for this variable" logic
- Prevent various NPEs when completing the first parameter of a
  parameterized type
  - eg.
    ```java
    List<| asdf = new ArrayList<>();
    ```
- move private check before deprecated check to avoid NPE trying to
  access the `IMethod` for a private JDK class member.
  (I think the root cause is that the private method doesn't exist in the
  stubbed jar, so we can't access the associated `IMethod`, but it does
  exist in javac's internal bindings, since javac has access to the real
  JDK class)
- when completing throws clause, add null checks when accessing the
  return type and name to compute the scanning index
  - constructors don't have a return type node, and an incomplete method
    declaration might not have a name
- replace a use of "Ljava/lang/Object;" with the constant from KeyUtils

Should fix 5 cases, and switch several test errors over to failures.

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
More error/NPE fixes in completion

- handle completing package-qualified method parameter type
- do not suggest overriding existing or declaring new methods in
  annotation type declarations
- when building override method completions, if the `IMethod` of a given
  method binding cannot be found, use the parameter names from
  the bindings instead
- prevent NPE when attempting to suggest completion for a method with
  an incomplete parameter type

Signed-off-by: David Thompson <davthomp@redhat.com>

Commit:
Fixes testParameterizedType_Bug83713 - Fix bad binding key

Signed-off-by: Rob Stryker <stryker@redhat.com>

rgrunber committed:
Add "since" 4.32 tag for Java 24 constants.

- Add api filter for JLS_Latest

Signed-off-by: Roland Grunberg <rgrunber@redhat.com>

datho7561 committed:
Improve first parameter completion for method

- If the method has no parameters and isn't a recovered "fake" binding,
  then perform the existing completion
  (suggest the method name,
  but don't change anything when the completion is applied)
- If the method binding has a parameter or is recovered,
  perform the default completion by suggesting accessible bindings.
  Ensure that statement-like keywords are filtered out
  - Add a test for this behaviour, since it's different from what
    happens upstream, and there are no existing tests that cover this
    behaviour

Signed-off-by: David Thompson <davthomp@redhat.com>

Signed-off-by: David Thompson <davthomp@redhat.com>

rgrunber committed:
Adjust based on javac library updates.

Signed-off-by: Roland Grunberg <rgrunber@redhat.com>
Co-Authored-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Skip running jdt.core tests without javac

- Don't bother building ECJ or jdt.core; grab those from the i-builds
- This will save 40-50 minutes in each Jenkins build
- We don't have substantial changes in jdt.core anymore.
  As such, we don't really need to test these changes.

Currently, there are 11 files that are changed:

PatternLocator.java
MatchingNodeSet.java (just whitespace changes)
MatchLocator.java
SourceIndexer.java
DOMToIndexVisitor.java
IJavaSearchDelegate.java (@since changes only)
DOMToModelPopulator.java
org.eclipse.jdt.internal.core.CompilationUnit (formatting changes only)
JavaCore.java (@since changes only)
AST.java (@since changes only)
CompletionEngine.java (make a few methods static)

As noted, only 6 of these files have "real" changes,
and of those, most of them are classes that we introduced upstream that
are only used when running with javac.

Signed-off-by: David Thompson <davthomp@redhat.com>

mickaelistria committed:
Set Java 24 requirement for org.eclipse.jdt.core.javac

datho7561 committed:
Revert some changes in jdt.core to match upstream

- Undo `@since` and cosmetic changes
- Remove changes from CompletionEngine

Signed-off-by: David Thompson <davthomp@redhat.com>

Commit:
[DOM Search] Fix testStringParameterizedMultipleArguments01 and 02, 03, 04 and others

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
Clean up / revert some unnecessary changes in jdt.core

Signed-off-by: Rob Stryker <stryker@redhat.com>

mickaelistria committed:
Fix some JAVADOC_LINE offsets

datho7561 committed:
Use recovered type for erasure in javac bindings

Should fix at least one test failure.

Signed-off-by: David Thompson <davthomp@redhat.com>

mickaelistria committed:
Filter visibility on search match (do not open type to get modifiers)

mickaelistria committed:
Avoid NPE when trying to resolve type from key

mickaelistria committed:
Some updates for Javadoc markdown conversion to Java 24

datho7561 committed:
Key for anonymous type declarations

- should fix one completion test

Signed-off-by: David Thompson <davthomp@redhat.com>

mickaelistria committed:
Prevent NPE

mickaelistria committed:
Map more problems

Fixes https://github.com/eclipse-jdtls/eclipse-jdt-core-incubator/issues/986
Fixes https://github.com/eclipse-jdtls/eclipse-jdt-core-incubator/issues/1184
Fixes https://github.com/eclipse-jdtls/eclipse-jdt-core-incubator/issues/1217
Fixes https://github.com/eclipse-jdtls/eclipse-jdt-core-incubator/issues/1233
Fixes https://github.com/eclipse-jdtls/eclipse-jdt-core-incubator/issues/1234
Fixes https://github.com/eclipse-jdtls/eclipse-jdt-core-incubator/issues/1250
Fixes https://github.com/eclipse-jdtls/eclipse-jdt-core-incubator/issues/1328

mickaelistria committed:
Use SignatureUtils everywhere in DOMCompletionEngine

datho7561 committed:
Fix intersection bindings

- Get key, name, qualified name, and toString of
  intersection bindings to match what ECJ has

This should turn some test errors into failures,
although we'll need to look into relevance numbers in order to get those
tests passing.

Signed-off-by: David Thompson <davthomp@redhat.com>

mickaelistria committed:
Fix insertion for enum constants

mickaelistria committed:
Fix relevance

* for enumConstants
* for types

datho7561 committed:
[WIP] Handle `IAccessRules` in completion

- hide types from other projects that are marked as inaccessible from completion
  (unless the appropriate settings to allow them to be completed are
  enabled)
- relevance numbers for those types

TODO:
- [ ] clean up the code, add comments, etc.

Signed-off-by: David Thompson <davthomp@redhat.com>

Commit:
Fix testLocalVariableDeclaration1 and 10+ others

Signed-off-by: Rob Stryker <stryker@redhat.com>

mickaelistria committed:
Improve variableName proposals for collections

mickaelistria committed:
Improvements for ExpectedTypes

mickaelistria committed:
Lazily resolve whether variable is unique

mickaelistria committed:
Don't rely on getKey to ensure bindings are unique

Instead make the binding be directly the key As 2 equals bindings in the
map. As long as  ashCode and equals are well implemented.

mickaelistria committed:
Stabilize some tests where ordering is important

Using LinkedHashSet instead of HashSet.

mickaelistria committed:
Fix location for MemberRef without qualifier

mickaelistria committed:
Better separate source and bin in sourcepath/classpath

Fixes https://github.com/eclipse-jdtls/eclipse-jdt-core-incubator/issues/1347

mickaelistria committed:
Complete on method name without expression

mickaelistria committed:
Do not fail on resolveBinding() when symbol has no matching .class

mickaelistria committed:
Always propose Object's methods

mickaelistria committed:
Improve perf for some completion based on index

Don't read modifiers from file, get them from TypeNameMatch

Commit:
Fix testStringNestedMultipleArguments02

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
[DOM Search] check current package for non-imported missing types

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
[DOM Search] Parameterized types are not having source range replaced when unnecessarily qualified

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
[DOM Search] Various issues trying to match unresolved patterns

Signed-off-by: Rob Stryker <stryker@redhat.com>

snjeza committed:
Incorrect hover and content assist with lombok

Signed-off-by: Snjezana Peco <snjezana.peco@redhat.com>

mickaelistria committed:
Share Names

As we often process the same projects, it seems better to keep a common
instance of Names, so that the names get instantiated only once.

Commit:
Fix testStringNestedMultipleArguments05

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
Fix testStringNestedSingleArgument01

Signed-off-by: Rob Stryker <stryker@redhat.com>

mickaelistria committed:
Completion Perf: Delay `scrapeAccessibleBindings`

scrapeAccessibleBindings may run some search and queries that can be
time consuming (~10% of completion time), however, the results of those
queries are not always useful.
So instead of forcing the computation of binding early, we allow to mark
them as "requested" so that they would be computed lazily, typically
when applying a method completion proposal.

datho7561 committed:
Mark array types referenced in method @see reference as used

- Fixes an import error when using javac with jdt.ui
- Prevent exception when DOM converting the following:

```java
/**
 * @see MyClass[]
 * @see MyClass[]#length()
 */
```

- In Javadoc @see, always visit the method parameters,
  even if the type is fully qualified

Signed-off-by: David Thompson <davthomp@redhat.com>

Commit:
Fix one issue - unsure which

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
Do not report imports for exact tests

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
Fix testParameterizedArrayTypeMultipleArguments01 and 6 others

Signed-off-by: Rob Stryker <stryker@redhat.com>

mickaelistria committed:
Don't use not-upstream DOMCodeSelector.isGenerated()

as try to use it can cause exception as long as it's not available

mickaelistria committed:
Ensure bindings are resolved before requesting import binding

snjeza committed:
Fix ASTConverterAST8Test.test0049()/test0052()

datho7561 committed:
Map another two error messages

Signed-off-by: David Thompson <davthomp@redhat.com>

Commit:
Fix several regressions in JavaSearchGenericTypeEquivalentTests

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
Reorg how validating type parameters works

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
[DOM Search Tests] - Fixes 8 tests, add some support for exact search on type arguments

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
[DOM Search] Fix 8 tests related to search queries with no type params when node qualifier has type params

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
Fix testStringParameterizedMultipleArguments01

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
Fix testStringParameterizedMultipleArguments02

Signed-off-by: Rob Stryker <stryker@redhat.com>

mickaelistria committed:
Cache and reuse symbols

Instead of always re-reading the jars for symbols, we create a local
cache that is capable of recreating the symbol into any context reliably
enough from memory.
This should drastically improve performance.

mickaelistria committed:
Fixes symbol cache

* ForAll/generic methods

datho7561 committed:
Completing static member no longer inserts class name twice

- Did some changes to how static classes are collected to fix related
  tests
- Did some changes to how completion text is built to fix related tests

Fixes #1369

Signed-off-by: David Thompson <davthomp@redhat.com>

Commit:
[DOM Search] Fixes 4

Signed-off-by: Rob Stryker <stryker@redhat.com>

mickaelistria committed:
Set -Xproc:none when APT is disabled

datho7561 committed:
Maybe fix semantic highlighting

There is an exception in the binding resolver,
this adds another check to hopefully prevent that.

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Fix some qualification relevances

- split up the additions of the relevances to make it a bit easier to debug
- enable "trim trailing whitespaces on save" in the javac fragment

Signed-off-by: David Thompson <davthomp@redhat.com>

Commit:
[DOM Search] Fixes several in JavaSearchGenericFieldTests

Signed-off-by: Rob Stryker <stryker@redhat.com>

datho7561 committed:
Convert `@serial` javadoc tag

Fixes #1363

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Completing enum sometimes double inserts the enum type name

Fix #1381

Signed-off-by: David Thompson <davthomp@redhat.com>

snjeza committed:
Fix JavaSearchBugsTests.testBug83804_Type and 3 others

Signed-off-by: Snjezana Peco <snjezana.peco@redhat.com>

mickaelistria committed:
Resolve method name to method

Commit:
Reformat all files in javac bundle to get rid of trailing spaces and other automatic changes

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
Fixes #1370 - moving static method via refactor fails, DOMSearch

Signed-off-by: Rob Stryker <stryker@redhat.com>

Fix regression

Signed-off-by: Rob Stryker <stryker@redhat.com>

Fix again

Signed-off-by: Rob Stryker <stryker@redhat.com>

Fix regressions

Signed-off-by: Rob Stryker <stryker@redhat.com>

datho7561 committed:
Prevent error when accepting keyword completion before loop

I think this error was just in the logs,
but it might also be a bit of a performance improvement.

Fixes #1367

Signed-off-by: David Thompson <davthomp@redhat.com>

mickaelistria committed:
Don't try to read symbol for JavacErrorMethodBinding.isVarargs

Hardcode value to false to prevent from trying to read the
(non-existing) MethodSymbol

datho7561 committed:
Run some jdt.ui tests

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 commented:
All the tests are failing due to something GTK-related. I'll dig into the jdt.ui test configuration to get a better idea of what's going on...

datho7561 commented:
Now there is an issue about the workbench not being initialized properly (but there are at least more passes this time).

datho7561 commented:
I was mixing JUnit 4 and 5 so I think that might be related

datho7561 commented:
Maybe it was because I didn't configure Tycho to use the UI properly?

mickaelistria commented:
Note that instead of Xvnc, you may be able to use wayland/mutter directly. FWIW, here is the bit of script to run in with GitHub actions https://unix.stackexchange.com/a/735032/560080 .
Can you please point to a particular log or error message about the workspace initialization issue?

datho7561 commented:
Here is one of the error messages: https://ci.eclipse.org/ls/job/jdt-core-incubator/job/PR-1393/2/testReport/RunQuickFixJavacTests%20org.eclipse.jdt.ui.tests.quickfix/AdvancedQuickAssistTest/testInverseVariable2b/

Many (though not all) of the QuickFix tests have this error.

I think this was fixed by adding the configuration to Tycho to run UI on a separate thread.

mickaelistria commented:
Looks like the tests like the one you pointed to and the workspace initialization have been working since introduction of `<useUIHarness>true</useUIHarness>`. Am I missing something?

datho7561 commented:
Nope, you're not missing anything, the tests were in fact working properly (until I introduced the commit trying to use Wayland).

I think I'll undo the commit trying to use Wayland, it seems like it won't be trivial to set up.

Commit:
Fixes testMethodDeclarationsElementPatternMultipleParamArguments02

Signed-off-by: Rob Stryker <stryker@redhat.com>

snjeza committed:
Fix ASTConverterTestAST8_2.test0602

Commit:
Fix regressions

Signed-off-by: Rob Stryker <stryker@redhat.com>

Commit:
Fixes testMethodReferencesStringPatternSingleParamArguments09 and others

Signed-off-by: Rob Stryker <stryker@redhat.com>

datho7561 committed:
Don't suggest keywords in an array initializer

eg. there should be no keywords when completing at the `|`

```java
int[] a = new int[] { | };
```

Signed-off-by: David Thompson <davthomp@redhat.com>

mickaelistria committed:
Small fixes for record bindings

mickaelistria committed:
Map more problems

Fixes https://github.com/eclipse-jdtls/eclipse-jdt-core-incubator/issues/1394
Fixes https://github.com/eclipse-jdtls/eclipse-jdt-core-incubator/issues/1395

mickaelistria committed:
Update GitHub action

mickaelistria committed:
Add workaround repo for repo.eclipse.org outage

mickaelistria committed:
Try archiving JUnit reports in GitHub actions

mickaelistria committed:
More GH JUnit reporting

mickaelistria committed:
More fixes to CachingClassSymbolClassReader

* Synchronize init of reflection
* Do not share the signatureBuffer, keep it instance specific to avoid
concurrent modification. Sharing doesn't bring any benefit anyway.

mickaelistria committed:
Only complete `default` inside annotations

mickaelistria committed:
Replace name instead of inserting when possible

mickaelistria committed:
Make ASTConverterMarkdownText more tolerant

A parser may decide to keep a TextElement as " blah" with (offset,
length)=(N,5); or as "blah" with (offset,length)=(N+1,4). Both are
valid.
Make test tolerate both.

mickaelistria committed:
Fix completion for `blah.|  unrelatedMethod()`

mickaelistria committed:
Search skips visiting EnumConstantDeclaration name

The parent EnumConstantDeclaration is already checked, no need to
re-check the name.

datho7561 committed:
Do not suggest keywords in empty block

I disagree personally but this is what JDT does.

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 committed:
Do not filter out types from file when completing field type

eg.

```java
public class MyLinkedList {

  public MyLinkedList(MyLinkedList next) {
    // ... assign next to a field
  }

  MyLink|
}
```

We were filtering `MyLinkedList` out because it was "default" completion,
but we were disabling default completion in this case.

Signed-off-by: David Thompson <davthomp@redhat.com>

datho7561 commented:
@robstryker @mickaelistria what are your thoughts on merging this? benefits are: better coverage of our bindings, search, and completion implementations, drawbacks are: more complex and slower test executions.

mickaelistria commented:
At the moment, I don't think we need more coverage. We still have ~3000 failures and adding new higher-level failures to that list isn't really going to help.
However, we can keep this PR open and rebase it from time to time to look at how UI tests are behaving.
